### ISSUE TITLE
The choice of the number of quantizers per subvector in IndexIVFPQ

### ISSUE THREAD

**Issue by minjiaz** [facebookresearch/faiss#259]
Hi there,

I'm running IndexIVFPQ and notice that when I set nbits (the number bits to index sub-codebook) to 16 (2 bytes), FAISS would throw an assertion failure saying that nbits should be equal or less than 8. This would put a restriction on the maximal number of codewords per subvector to 256. 

My understanding of PQ is that increasing nbits would significantly increase the number of codewords that can be represented without increasing memory footprint too much. This is because the total number of codewords is (2^nbits)^m (where m is the number of subvectors), the codebook size is (i.e., m * (2^nbits)), and the number of bytes per vector is (m * nbits / 8). Hence, increasing nbits will likely yield to a better memory-recall tradeoff. 

The place where the exception is thrown is at the following place:
Error in faiss::IndexIVFPQ::IndexIVFPQ(faiss::Index*, size_t, size_t, size_t, size_t) at IndexIVFPQ.cpp:50: Error: 'nbits_per_idx <= 8' failed. 
Also, in Figure 6 of the paper "product quantization for nearest neighbor search", K* is set to 4096, which is equivalent to nbits set to 12. I wonder if there are any particular reasons that in the current implementation nbits is limited to 8, and what I can do to increase nbits to larger values such as 16?

**Comment by wickedfoo** [comment]
The core ProductQuantizer.h/.cpp that calculates the codes supports 2 bytes per code, but both the CPU and GPU IVFPQ index which uses the ProductQuantizer do not, mainly for practical reasons. 

The CPU version could be fixed, but one consideration is that at, say, nbits == 16 (65536 codes/subvector), you rapidly get into a situation where the codebook would no longer fit into L1/L2 and maybe not even the last level cache (L3), so it becomes impractical to use. The GPU code cannot support this many codes per subvector, maybe only 9/10 bits per code at most, due to shared memory limitations (48 KiB), but this would place further restrictions on how many subvectors you are using, etc.

**Comment by minjiaz** [comment]
Hi wickedfoo, 

Thank you very much for your reply! It is good to know that nbits <= 8 can be fixed on the CPU version. Can you please let me what changes roughly need to be made in order to enable nbits = 16 if the required changes are not too many? 

I see your point on restricting codebook size to let it fit in fast cache. But I think even with nbits == 16, there is still a large chance that the codebook can fit into L3. My understanding is that nbits == 16 means each sub-codeword takes 16-bit (2-byte). Assume there are 65536 sub-codeword, the sub-codebook size is 2 X 65536 = 128KB. Given L3 cache is 64MB on most Intel Xeon CPU machines today, that supports an M (the number of subvectors) potentially as large as 500 for the whole codebook to host in L3. However, by increasing nbits, we will having more codewords even with some reduction on M. Hence, with some loss of execution time, recall might be improved.  This would benefit scenarios where memory space and recall are considered more important and allow some sacrifice on execution time.

**Comment by wickedfoo** [comment]
The factor "2" is only the size of the codes for the encoded vectors themselves, not the size of the codebook that is used to decode it.

The codebook itself is sizeof(float) x number of sub-dimensions x number of codes; i.e., the centroids of the residual for each sub-dimension. So a typical value would be more like 4 * 8 * 65536 = 2097152 bytes or more.

At query time, a distance is calculated against each IVF cluster centroid, so this data structure that is used during query is something like sizeof(float) x number of sub-vectors x number of IVF centroids, which is comparable to the above. This is something that will trash the cache as well.

Depending upon the number of sub-vectors chosen, at some point an index like IVFFlat will likely perform better, as it has more regular memory access patterns and will make better use of the bandwidth provided by caches than the random accesses performed by codeword lookup. I take your point that an encoding like 4 x 8192 or 8 x 16384 could work well and is another possible operating point versus other partitioning points, but many of these points can't happen on the GPU (4 x 8192 could, but larger in either dimension can't).

For fixing the CPU side, one would have to look at IndexIVFPQ.cpp and dependencies and make sure that codes are not treated as exclusively bytes. This is not a priority of ours now.

**Comment by minjiaz** [comment]
Hi Jeff,

Thanks for your detailed reply. That makes sense. I think the total size of the codebook is:
(input vector dimension/ m) * sizeof(float) * (number of sub-dimension) * (number of sub-codeword each in sub-dimension). 

It also makes sense that the pre-computation of the distance look-up table (distance between subvectors of a query and centroids of sub-dimensions) can pollute the cache.  

I agree that for GPU, there is a practical limitation on the codebook size due to the limited shared memory size. However, I still think on CPU the codebook size can be a lot larger as long as it can sit in L3 cache. There are studies that have shown with larger nbit (e.g., 16), it offers a larger speed-accuracy trade-off space (as shown in table 1 in "Accelerated Nearest Neighbor Search withick ADC" ICMR'17).

Another reason for hosting the codebook in L3: In the precomputation phase, the codebook is used for distance look-up table computation. Once the precomputation phase is done, in the scanning phase each data point requires two memory loads to estimate its distance to the query: one from memory to load the PQ code (indexes to codewords), and one from decoding by checking the distance lookup table. Since there is a main memory load for each data point anyway, it is likely that the latency gain is diminishing by hosting the codebook in L2 vs. L3.

**Comment by mdouze** [comment]
It seems the question is answered, closing.
